---
title: "Plot GENCODE stats"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(reshape2)
library(rtracklayer)
library(scales)
library(cowplot)
theme_set(theme_bw())
knitr::opts_chunk$set(echo = TRUE)
```

## Genomic region stats

Plot.

```{r load_and_plot, warning=FALSE, message=FALSE}
stat_files <- list.files(path = ".", pattern = "stats$")
file_list <- lapply(stat_files, function(x){
  read_delim(file = x, delim = " ", col_names = c("var", "value"))
})

names(file_list) <- sub(pattern = "*.annotation.gtf.stats", replacement = "", x = stat_files)

my_df <- as.data.frame(do.call(rbind, sapply(file_list, function(x){
  x[, 2]
})))

colnames(my_df) <- sub(pattern = ":", replacement = "", x = file_list[[1]]$var)
my_df$version <- sub(pattern = "gencode.v(\\d\\d).value", replacement = "\\1", x = row.names(my_df))
row.names(my_df) <- NULL
my_df <- melt(my_df, id.vars = "version")

my_df %>%
  filter(str_detect(variable, "coverage")) %>%
  ggplot(., aes(version, value, fill = variable)) +
  geom_col() +
  labs(x = "GENCODE version", y = "Percentage", title = "Genomic region proportions") -> p1

my_df %>%
  filter(str_detect(variable, "length")) %>%
  ggplot(., aes(version, value, fill = variable)) +
  geom_col() +
  labs(x = "GENCODE version", y = "Average length (bp)", title = "Genomic region lengths") -> p2

p1
p2
```

Save plots.

```{r save_plot, warning=FALSE, message=FALSE}
ggsave(filename = "genomic_region_proportion.png", plot = p1)
ggsave(filename = "genomic_region_length.png", plot = p2)
```

## GTF stats

Load all GTF files into R.

```{r load_gtf, warning=FALSE, message=FALSE}
gtf_files <- list.files(path = ".", pattern = "gtf.gz$")

gtf_obj <- lapply(gtf_files, import)

names(gtf_obj) <- sub(pattern = "*.annotation.gtf.gz", replacement = "", x = gtf_files)
```

Plot `gene_type`.

```{r plot_gene_type, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
plot_gene_type <- function(gr, plot_title){
  gr %>%
    as.data.frame() %>%
    filter(type == "gene") %>%
    group_by(gene_type) %>%
    summarise(count = n()) %>%
    arrange(desc(count)) %>%
    mutate(gene_type = factor(x = gene_type, levels = gene_type)) %>%
    ggplot(., aes(gene_type, y = count)) +
      geom_col() +
      scale_y_log10() +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.x = element_blank()) +
      scale_y_continuous(labels = comma) +
      labs(title = plot_title)
}

p1 <- plot_gene_type(gtf_obj$gencode.v20, "GENCODE version 20")
p2 <- plot_gene_type(gtf_obj$gencode.v30, "GENCODE version 30")
p3 <- plot_gene_type(gtf_obj$gencode.v31, "GENCODE version 31")
p4 <- plot_gene_type(gtf_obj$gencode.v35, "GENCODE version 35")

all_plot <- plot_grid(plotlist = list(p1, p2, p3, p4))
all_plot
ggsave(filename = "gene_type.png", plot = all_plot)
```

Plot `transcript_support_level`, which are [transcript scores](https://www.gencodegenes.org/pages/data_format.html) according to how well mRNA and EST alignments match over its full length:

* 1 (all splice junctions of the transcript are supported by at least one non-suspect mRNA),
* 2 (the best supporting mRNA is flagged as suspect or the support is from multiple ESTs),
* 3 (the only support is from a single EST),
* 4 (the best supporting EST is flagged as suspect),
* 5 (no single transcript supports the model structure),
* NA (the transcript was not analyzed)

```{r plot_transcript_support, message=FALSE, warning=FALSE, fig.width=8, fig.height=6}
plot_transcript_support <- function(gr, plot_title){
  gr %>%
    as.data.frame() %>%
    filter(type == "transcript") %>%
    group_by(transcript_support_level) %>%
    summarise(count = n()) %>%
    arrange(desc(count)) %>%
    filter(!is.na(transcript_support_level)) %>%
    mutate(transcript_support_level = factor(x = transcript_support_level, levels = c(1:5, "NA"))) %>%
    ggplot(., aes(transcript_support_level, y = count)) +
      geom_col() +
      #scale_y_log10() +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.x = element_blank()) +
      scale_y_continuous(labels = comma) +
      labs(title = plot_title)
}

p1 <- plot_transcript_support(gtf_obj$gencode.v21, "GENCODE version 21")
p2 <- plot_transcript_support(gtf_obj$gencode.v30, "GENCODE version 30")
p3 <- plot_transcript_support(gtf_obj$gencode.v33, "GENCODE version 33")
p4 <- plot_transcript_support(gtf_obj$gencode.v35, "GENCODE version 35")

all_plot <- plot_grid(plotlist = list(p1, p2, p3, p4))
all_plot
ggsave(filename = "transcript_support_level.png", plot = all_plot)
```
